esmf_regrid/schemes.py

```diff
@@ -1,19 +1,315 @@
-"""TBD: public module docstring."""
+"""Provides an iris interface for regridding."""
+
+import copy
+
+import iris
+from iris.analysis._interpolation import get_xy_dim_coords
+import numpy as np
+
+from esmf_regrid.esmf_regridder import GridInfo, Regridder
 
 __all__ = [
     "ESMFAreaWeighted",
+    "ESMFAreaWeightedRegridder",
+    "regrid_rectilinear_to_rectilinear",
 ]
 
 
+def _bounds_cf_to_simple_1d(cf_bounds):
+    assert (cf_bounds[1:, 0] == cf_bounds[:-1, 1]).all()
+    simple_bounds = np.empty((cf_bounds.shape[0] + 1,), dtype=np.float64)
+    simple_bounds[:-1] = cf_bounds[:, 0]
+    simple_bounds[-1] = cf_bounds[-1, 1]
+    return simple_bounds
+
+
+def _cube_to_GridInfo(cube):
+    # This is a simplified version of an equivalent function/method in PR #26.
+    # It is anticipated that this function will be replaced by the one in PR #26.
+    #
+    # Returns a GridInfo object describing the horizontal grid of the cube.
+    # This may be inherited from code written for the rectilinear regridding scheme.
+    lon = cube.coord("longitude")
+    lat = cube.coord("latitude")
+    # Ensure coords come from a proper grid.
+    assert isinstance(lon, iris.coords.DimCoord)
+    assert isinstance(lat, iris.coords.DimCoord)
+    # TODO: accommodate other x/y coords.
+    # TODO: perform checks on lat/lon.
+    #  Checks may cover units, coord systems (e.g. rotated pole), contiguous bounds.
+    if cube.coord_system() is None:
+        crs = None
+    else:
+        crs = cube.coord_system().as_cartopy_crs()
+    return GridInfo(
+        lon.points,
+        lat.points,
+        _bounds_cf_to_simple_1d(lon.bounds),
+        _bounds_cf_to_simple_1d(lat.bounds),
+        crs=crs,
+        circular=lon.circular,
+    )
+
+
+def _regrid_along_grid_dims(regridder, data, grid_x_dim, grid_y_dim, mdtol):
+    data = np.moveaxis(data, [grid_x_dim, grid_y_dim], [-1, -2])
+    result = regridder.regrid(data, mdtol=mdtol)
```

Comment:
This seems to be a different interface to the `regridder(data)` below.  Is the `regridder` argument a `Regridder` or an `ESMFWeightedAreaRegridder`? This might be the perfect opportunity to add some type hints!

Reply:
Incidentally, why do these two objects have different interfaces? From the names you'd intuitively expect one to be a subclass of the other, but that is not the case.

Reply:
So the `ESMFWeightedAreaRegridder` deals with iris objects and `Regridder` deals with `GridInfo`/`MeshInfo` objects. For this reason, I don't think it's appropriate to have one be a subclass of the other. `Regridder` is more generic while `ESMFWeightedAreaRegridder` is dealing with more specific cases. There is probably a case for the names to be more thought out. I believe we came up with these names a while ago and at the time the name `ESMFWeightedAreaRegridder` suggested that what would be specific about this regridding scheme is that it would use ESMF and be an instance of area weighted regridding. I suppose the idea was that eventually our back end would be able to handle more kinds of regridding and potentially use different packages to do regridding with. At the present, we don't have any other kinds of regridding so this isn't really what sets this appart from `Regridder`, though this could still change in the future.